Meter.c

```diff
@@ -277,50 +428,166 @@ static void GraphMeterMode_draw(Meter* this, int x, int y, int w) {
+            for (int i = 0; i < items; i++) {
+               stack1[i + 1] = stack1[i + 1] * this->full / stack1[items + 1];
+            }
+         }
+         scale = this->full;
+      } else {
+         int exp;
+         (void) frexp(MAXIMUM(stack1[items], stack2[items]), &exp);
+         scale = ldexp(1.0, exp);
+      }
+
+      // Determine the dominant color per cell in the graph.
+      // O(GRAPH_HEIGHT + items) (linear time)
+      for (int step = 1; ; step <<= 1) {
+         int stack1Start = 0, stack2Start = 0;
+         double low, high = 0.0;
+         for (int h = 0; ; h += step) {
+            size_t offset = (data->colorRowSize <= GRAPH_HEIGHT) ? h :
+                            (h * 2) + step;
+            if (offset >= data->colorRowSize)
+               break;
+
+            low = high;
+            high = scale * (h + step) / GRAPH_HEIGHT;
+            assert(low == scale * (h) / GRAPH_HEIGHT);
+
+            double maxArea = 0.0;
+            int color = BAR_SHADOW;
+            double maxArea2 = 0.0;
+            int color2 = BAR_SHADOW;
+            for (int i = MINIMUM(stack1Start, stack2Start); ; i++) {
+               if (stack1[i] < high)
+                  stack1Start = i;
+               if (stack2[i] < high)
+                  stack2Start = i;
+               if (i >= items)
+                  break; // No more items
+               if (stack1[i] >= high && stack2[i] >= high) {
+                  // This cell is finished. Rest of values are out of this.
+                  break;
+               }
+               // Skip items that have no area in this cell.
+               if (stack1[i] >= high)
+                  i = MAXIMUM(i, stack2Start);
+               if (stack2[i] >= high)
+                  i = MAXIMUM(i, stack1Start);
+               double area;
+               area = CLAMP(stack1[i + 1], low, high) +
+                      CLAMP(stack2[i + 1], low, high);
+               area -= (CLAMP(stack1[i], low, high) +
+                        CLAMP(stack2[i], low, high));
+               if (area > maxArea) {
+                  maxArea = area;
+                  color = Meter_attributes(this)[i];
+               }
+               if (data->drawOffset > 0) {
+                  area = CLAMP(stack2[i + 1], low, high);
+                  area -= CLAMP(stack2[i], low, high);
+                  if (area > maxArea2) {
+                     maxArea2 = area;
+                     color2 = Meter_attributes(this)[i];
+                  }
+               }
+            }
+            data->colors[(GRAPH_NUM_RECORDS - 2) * data->colorRowSize +
+                         offset] = color;
+            if (data->drawOffset > 0) {
+               data->colors[(GRAPH_NUM_RECORDS - 1) * data->colorRowSize + offset] = color2;
+            }
+         }
+         if (data->colorRowSize <= GRAPH_HEIGHT) {
+            break;
+         } else if (step >= 2 * GRAPH_HEIGHT) {
+            break;
+         }
+      }
```

Comment:
Short, highlevel description would be fine. Also this looks awfully complicated and should probably be split into its own function.

Also some blank lines for better visual grouping might help.